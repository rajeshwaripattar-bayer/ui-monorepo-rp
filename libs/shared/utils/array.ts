import map from 'lodash/map'
import fill from 'lodash/fill'

export const range = (start: number, stop: number, step: number = 1) =>
  map(fill(Array(Math.ceil((stop - start) / step)), start), (x, y) => x + y * step)
