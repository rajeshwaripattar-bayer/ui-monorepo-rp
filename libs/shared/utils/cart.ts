import { ProductWithPrice } from '@gc/components/types'
import { Entry } from '@gc/types'

const entryMatcher = (cartEntry: Entry, entry: Entry, skipLocationCheck = false) =>
  cartEntry.product.code === entry.product.code &&
  cartEntry.cropCode === entry.cropCode &&
  (cartEntry.storageLocation?.locationCode === entry.storageLocation?.locationCode || skipLocationCheck) &&
  !cartEntry.rejected

export const getMatchingCartEntry = (entries: Entry[] = [], entry: Entry, skipLocationCheck = false) =>
  entries.find((cartEntry) => entryMatcher(cartEntry, entry, skipLocationCheck))

export const getMatchingCartEntryIndex = (entries: Entry[] = [], entry: Entry, skipLocationCheck = false) =>
  entries.findIndex((cartEntry) => entryMatcher(cartEntry, entry, skipLocationCheck))

const entryMatcherUsingProduct = (cartEntry: Entry, product: ProductWithPrice, skipLocationCheck = false) =>
  cartEntry.product.code === product.code &&
  (cartEntry.storageLocation?.locationCode === product.warehouse?.value || skipLocationCheck) &&
  !cartEntry.rejected

export const getMatchingCartEntryUsingProduct = (
  entries: Entry[] = [],
  product: ProductWithPrice,
  skipLocationCheck = false
) => entries.find((cartEntry) => entryMatcherUsingProduct(cartEntry, product, skipLocationCheck))

export const getMatchingCartEntryIndexUsingProduct = (
  entries: Entry[] = [],
  product: ProductWithPrice,
  skipLocationCheck = false
) => entries.findIndex((cartEntry) => entryMatcherUsingProduct(cartEntry, product, skipLocationCheck))
