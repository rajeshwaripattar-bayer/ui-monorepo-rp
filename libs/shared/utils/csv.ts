import _ from 'lodash'
import { IColumn, Datas } from 'react-csv-downloader/dist/esm/lib/csv'
import { FarmerDetails, CropZone, TableColConfig } from '@gc/types'
import { HeaderType } from '../components/src/ui-common/table/Table'
import { utils, writeFile, write } from 'xlsx'

export function getFileNameWithDate(fileName: string) {
  const today = new Date()
  const dd = String(today.getDate()).padStart(2, '0')
  const mm = String(today.getMonth() + 1).padStart(2, '0') // January is 01
  const yyyy = today.getFullYear()
  return fileName + yyyy + mm + dd
}

export function getCSVColumns<T>(colData: HeaderType<T>[] | TableColConfig[]) {
  const exportedCols: IColumn[] = []
  colData?.forEach((e) => {
    if (e.excludeFromDownload) return
    if (e.accessor !== undefined && e.header !== undefined) {
      exportedCols.push({
        id: e.accessor.toString(),
        displayName: e.header
      })
    }
  })
  return exportedCols
}

export function formatFarmerDetailsCSVData(farmerDetails: FarmerDetails[]) {
  let clonedFarmerDetails = _.cloneDeep(farmerDetails)
  return clonedFarmerDetails.map((farmerDetail) => {
    // convert if any field value is number to string
    Object.entries(farmerDetail).forEach(([key, value]) => {
      if (typeof value === 'number' && key in farmerDetail) {
        ;(farmerDetail as any)[key] = (value as number).toString()
      }
    })
    // convert cropZones to individual columns
    return farmerDetail.cropZones?.reduce(
      (accum: object, cz: CropZone) => {
        Object.assign(accum, {
          [`${cz.crop}CyZone`]: cz.cyZone,
          [`${cz.crop}CyReassigned`]: cz.cyReassigned ? cz.cyReassigned : '',
          [`${cz.crop}PyZone`]: cz.pyZone,
          [`${cz.crop}PyReassigned`]: cz.pyReassigned ? cz.pyReassigned : ''
        })
        return accum
      },
      {
        ..._.omit(farmerDetail, ['cropZones', 'cyOrder']),
        cyOrder:
          farmerDetail.cyOrder === null || farmerDetail.cyOrder === undefined
            ? farmerDetail.cyOrder
            : farmerDetail.cyOrder.toString()
      }
    )
  }) as Datas
}

export function downloadXlsx(
  columns: { displayName: string; id: string }[],
  data: Record<string, any>[],
  fileName: string
) {
  // Create the header row using displayName
  const headers = columns.map((col) => col.displayName)
  // Convert data to a 2D array using the id from columnDefs
  const dataArray = data.map((row) => columns.map((col) => row[col.id]))
  // Prepend the headers to the data array
  const finalDataArray = [headers, ...dataArray]
  // Create a workbook
  const workbook = utils.book_new()
  // Create a worksheet
  const worksheet = utils.aoa_to_sheet(finalDataArray) // Use aoa_to_sheet for array of arrays
  // Append the worksheet to the workbook
  utils.book_append_sheet(workbook, worksheet, fileName)
  // type to xlsx
  write(workbook, { bookType: 'xlsx', type: 'binary' })
  // Write the workbook to a xlsx file
  writeFile(workbook, `${getFileNameWithDate(fileName)}.xlsx`)
}
