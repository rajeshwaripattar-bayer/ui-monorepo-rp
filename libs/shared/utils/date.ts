import moment from 'moment'
import { Locale } from '@gc/types'

export function calculateOffsetDate(timePeriod: string, duration: number) {
  const date = new Date()
  switch (timePeriod) {
    case 'month':
      date.setMonth(date.getMonth() + duration)
      break
    case 'week':
      date.setDate(date.getDate() + 7 * duration)
      break
    default:
      date.setDate(date.getDate() + duration)
  }
  return date
}

export function formatDateWithTimezoneOffset(inputDate?: Date) {
  return inputDate ? moment(inputDate).format('yyyy-MM-DDTHH:mm:ssZZ') : ''
}

export function getLongDate(date: Date, locale: Locale): string {
  return date.toLocaleDateString(locale.code, {
    month: 'long',
    day: '2-digit',
    year: 'numeric'
  })
}

export function getDateFromUTC(date: Date, locale: Locale): string {
  return date ? date.toLocaleDateString(locale.code, { year: '2-digit', month: '2-digit', day: '2-digit' }) : ''
}
