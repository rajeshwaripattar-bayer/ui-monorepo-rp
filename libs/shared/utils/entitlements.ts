export const hasEntitlement = (entitlements: string[], partial: string) => {
  return entitlements.some((e) => e.includes(partial))
}

export const hasAnyEntitlement = (entitlements: string[], partials: string[] = []) => {
  return partials.some((partial) => hasEntitlement(entitlements, partial))
}

export const hasNbmEntitlement = (entitlements: string[], lob: string) => {
  return hasEntitlement(entitlements, `${lob}:nbm:`)
}

export const hasNbmProgramEntitlement = (entitlements: string[], lob: string, programName: string) => {
  return hasEntitlement(entitlements, `${lob}:nbm:${programName}`)
}
