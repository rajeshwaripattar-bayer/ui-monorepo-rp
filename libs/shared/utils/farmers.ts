import { CropZone, DomainDef, FarmerDetails } from '@gc/types'
import { fetchStore } from './module'

export const getAccountName = (farmerDetail: FarmerDetails) => {
  return farmerDetail?.farmName || farmerDetail?.growerName
}

export const checkFarmerDetails = (farmerDetails: FarmerDetails) => {
  const requiredFields = [
    'growerSapId',
    'growerIrdId',
    'streetAddress',
    'city',
    'state',
    'zipCode',
    'phoneNumber',
    'email'
  ]
  return requiredFields.every((key: string) => farmerDetails[key as keyof FarmerDetails])
}

export const filterCropZone = (cropZones: CropZone[]) => {
  const {
    gcPortalConfig: { crops }
  } = fetchStore('domainDef') as DomainDef
  return crops ? cropZones?.filter((cz) => crops.includes(cz.crop?.toLowerCase())) : cropZones
}
