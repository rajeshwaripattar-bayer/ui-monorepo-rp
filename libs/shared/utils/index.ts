export * from './array'
export * from './cart'
export * from './test-utils'
export * from './service'
export * from './format'
export * from './media'
export * from './date'
export * from './csv'
export * from './date'
export * from './search'
export * from './module'
export * from './programs'
export * from './entitlements'
export * from './farmers'
export * from './filter'
export * from './products'
export * from './format'
export * from './mapper'
export * from './media'
export * from './module'
export * from './search'
export * from './service'
export * from './sort'
export * from './test-utils'
export * from './isDev'
