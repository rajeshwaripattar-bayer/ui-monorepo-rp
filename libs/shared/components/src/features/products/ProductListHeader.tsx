import { IconButtonProps } from '@element/react-icon-button'
import { resolutions } from '@gc/constants'
import { useLocale, useScreenRes } from '@gc/hooks'
import { getCurrencyFormat } from '@gc/utils'
import { useTranslation } from 'react-i18next'
import Band from '../../ui-common/band/Band'

/* eslint-disable-next-line */
export interface ProductListHeaderProps {
  icon?: string | React.ReactNode
  count: number
  crop: string
  averagePrice?: number
  currencyIso?: string
  containerClassName?: string
  trailingIconButtonProps?: IconButtonProps
}

export function ProductListHeader({
  count,
  crop,
  icon,
  averagePrice,
  currencyIso,
  trailingIconButtonProps
}: ProductListHeaderProps) {
  const { t } = useTranslation()
  const locale = useLocale()
  const res = useScreenRes()

  const isDesktop = res > resolutions.M1023
  const averagePriceLabel = `${
    isDesktop ? `${t('common.average_desktop.label')}:` : t('common.average_mobile.label')
  } ${getCurrencyFormat(currencyIso || 'USD', averagePrice || 0, locale)}/${t('common.ssu.label')}`
  return (
    <Band
      placement='list'
      iconProps={icon ? { icon } : undefined}
      primaryText={`${count} ${crop} ${t('common.product.label', { count })}`}
      secondaryText={averagePrice ? averagePriceLabel : ''}
      {...(trailingIconButtonProps ? { trailingIconButtonProps: [trailingIconButtonProps] } : {})}
    />
  )
}

export default ProductListHeader
