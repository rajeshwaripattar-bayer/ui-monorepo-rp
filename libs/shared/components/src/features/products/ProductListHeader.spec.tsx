import { render, screen } from '@gc/utils'

import ProductListHeader from './ProductListHeader'

describe('ProductListHeader', () => {
  it('should render successfully for desktop', () => {
    const wrapper = render(
      <ProductListHeader icon='tune' count={2} crop='Corn' averagePrice={100} currencyIso='USD' />,
      { width: 1440 }
    )
    expect(screen.queryByText(/common.average_desktop.label/)).toBeTruthy()
    expect(screen.queryByText(/subtitle2/)).toBeDefined()
    expect(wrapper).toMatchSnapshot()
  })

  it('should render successfully for mobile', () => {
    const wrapper = render(
      <ProductListHeader icon='tune' count={2} crop='Corn' averagePrice={100} currencyIso='USD' />,
      { width: 599 }
    )
    expect(screen.queryByText(/common.average_mobile.label/)).toBeTruthy()
    expect(screen.queryByText(/subtitle1/)).toBeDefined()
    expect(wrapper).toMatchSnapshot()
  })
})
