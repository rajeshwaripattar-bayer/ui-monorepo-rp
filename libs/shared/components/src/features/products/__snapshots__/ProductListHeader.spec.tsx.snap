// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductListHeader should render successfully for desktop 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container undefined"
      >
        <i
          aria-hidden="false"
          class="lmnt lmnt-icon material-icons lmnt-icon--medium lmnt-icon--fill lmnt-icon--fill-secondary"
        >
          tune
        </i>
        <div
          class="info"
        >
          <span
            class="mdc-typography lmnt mdc-typography--subtitle2"
          >
            2
             
            Corn
             
            common.product.label
          </span>
          <span
            class="mdc-typography lmnt mdc-typography--caption"
          >
            common.average_desktop.label:
             
            $100.00
            /
            common.unit.label
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container undefined"
    >
      <i
        aria-hidden="false"
        class="lmnt lmnt-icon material-icons lmnt-icon--medium lmnt-icon--fill lmnt-icon--fill-secondary"
      >
        tune
      </i>
      <div
        class="info"
      >
        <span
          class="mdc-typography lmnt mdc-typography--subtitle2"
        >
          2
           
          Corn
           
          common.product.label
        </span>
        <span
          class="mdc-typography lmnt mdc-typography--caption"
        >
          common.average_desktop.label:
           
          $100.00
          /
          common.unit.label
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ProductListHeader should render successfully for mobile 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container undefined"
      >
        <i
          aria-hidden="false"
          class="lmnt lmnt-icon material-icons lmnt-icon--medium lmnt-icon--fill lmnt-icon--fill-secondary"
        >
          tune
        </i>
        <div
          class="info"
        >
          <span
            class="mdc-typography lmnt mdc-typography--subtitle1"
          >
            2
             
            Corn
             
            common.product.label
          </span>
          <span
            class="mdc-typography lmnt mdc-typography--caption"
          >
            common.average_mobile.label
             
            $100.00
            /
            common.unit.label
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container undefined"
    >
      <i
        aria-hidden="false"
        class="lmnt lmnt-icon material-icons lmnt-icon--medium lmnt-icon--fill lmnt-icon--fill-secondary"
      >
        tune
      </i>
      <div
        class="info"
      >
        <span
          class="mdc-typography lmnt mdc-typography--subtitle1"
        >
          2
           
          Corn
           
          common.product.label
        </span>
        <span
          class="mdc-typography lmnt mdc-typography--caption"
        >
          common.average_mobile.label
           
          $100.00
          /
          common.unit.label
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
